[Sema] Warn for redundant access-level modifiers on setters or used in an extension.

dingobye opened the pull request:
~~This patch adds warning for redundant access-level modifiers used in an extension.~~

This patch adds warnings for redundant access-level modifiers for two aspects: (1) used on members declared in an extension, and (2) used on property setters.

(1) For the following example, it produces a warning - _**'public' modifier is redundant for instance method declared in a public extension**_.
```diff
 public extension SomeClass {
     public func foo() {}
+    ^~~~~~~
 }
```

(2) For instance below, it produces a warning - _**'public(set)' modifier is redundant for a public property**_.
```diff
 public SomeClass {
     public public(set) var x: Int = 0
+           ^~~~~~~~~~~~
 }
```

In addition, it refines the diagnostics of `access_control_ext_member_more` issues, in case the fixit
could suggest redundant modifiers. Consider the example below, the existing fixit suggests replacing `public` by `internal`, which is indeed a redundant modifier. This patch changes the fixit from replacement to removal.
```diff
 internal extension SomeClass {
     public func foo() {}
     ^~~~~~
-    internal
 }
```

Moreover, the corresponding redundancy warnings in the existing code of Swift projects are also fixed. These include:
- [x] stdlib in the main Swift repo as fixed by this patch
- [x] swift-corelibs-foundation repo [#1663](https://github.com/apple/swift-corelibs-foundation/pull/1663)
- [x] swift-corelibs-libdispatch repo [#390](https://github.com/apple/swift-corelibs-libdispatch/pull/390)
- [x] swift-corelibs-xctest repo [#218](https://github.com/apple/swift-corelibs-xctest/pull/218)
- [x] swiftpm repo [#1746](https://github.com/apple/swift-package-manager/pull/1746)

Resolves [SR-8453](https://bugs.swift.org/browse/SR-8453).

dingobye committed:
[Sema] Warn when redundant access-level modifier is added in an extension.

This patch adds warning for redundant access-level modifiers
used in an extension. It also refines the diagnostics of
access_control_ext_member_more issues, in case the fixit
could suggest redundant modifiers.

Resolves: SR-8453.

dingobye commented:
@jrose-apple Hi Jordan, would you help review this, please?

jrose-apple commented on test/SILGen/accessibility_warnings.swift:
```diff
@@ -56,15 +56,9 @@ extension PrivateStruct {
-  public func extMemberPublic() {}
-  // CHECK-DAG: sil private @$S22accessibility_warnings12PublicStructV07extImplC033_5D2F2E026754A901C0FF90C404896D02LLyyF
-  private func extImplPublic() {}
-}
```
Please don't remove this; just add the expected diagnostics. It's still checking the SIL-level linkage.

dingobye replied:
Fixed.

jrose-apple commented on test/Constraints/tuple_arguments.swift:
```diff
@@ -1632,12 +1632,14 @@ do {
     }
   }
 
+  // expected-warning@+1 {{'public' modifier is redundant for instance method declared in a public extension}}
```
Since this test isn't really about the access modifiers, it's probably better to just remove the `public` from the extension (both here and in test/Compatibility).

dingobye replied:
Done.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1189,7 +1189,12 @@ ERROR(access_control_setter_more,none,
 WARNING(access_control_ext_member_more,none,
     "declaring %select{%error|a fileprivate|an internal|a public|open}0 %1 in "
-    "%select{a private|a fileprivate|an internal|public|%error}2 extension",
+    "%select{a private|a fileprivate|an internal|a public|%error}2 extension",
```
Thanks for fixing this. :-)

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -3997,9 +3998,14 @@ void swift::fixItAccess(InFlightDiagnostic &diag, ValueDecl *VD,
-      diag.fixItReplace(attr->getLocation(), fixItString.drop_back());
+      if (shouldNotReplace) {
+        // Remove the attribute if replacement is not preferred.
+        diag.fixItRemove(attr->getRange());
```
Can you add a test with `public(set)`?

dingobye replied:
Good catch! I extended the solution to also produce warnings for redundant access modifiers on setters.

jrose-apple commented on lib/Sema/MiscDiagnostics.h:
```diff
@@ -51,8 +51,11 @@ void performTopLevelDeclDiagnostics(TypeChecker &TC, TopLevelCodeDecl *TLCD);
+                 ValueDecl *VD,
+                 AccessLevel desiredAccess,
+                 bool isForSetter = false,
+                 bool shouldNotReplace = false);
```
Nitpick: "shouldNotReplace" doesn't tell us what to do instead. Maybe "shouldUseDefaultAccess" or something?

dingobye replied:
Yeah, `shouldUseDefaultAccess ` sounds better!

jrose-apple reviewed (commented):
Thanks for picking this up, Ding! I have a few comments but the general approach seems reasonable.

dingobye committed:
Make it more comprehensive to warn when redundant access
modifier is used in an extension. In addition, add warnings
for access modifier redundancy on property setters; and
address comments from Jordan Rose.

dingobye commented:
Thank you for your feedback, Jordan! I updated the code as you suggested.

In addition, this update produces a warning for the following case:

_**'fileprivate' modifier is redundant for instance method declared in a private (equivalent to fileprivate)**_
```diff
 private extension SomeClass {
     fileprivate func foo() {}
+    ^~~~~~~~~~~~
 }
```
My earlier solution didn't report any warning for such cases. I think it more appropriate to warn, since it makes the entire redundancy checking comprehensive and consistent.

jrose-apple commented:
@swift-ci Please test

dingobye commented:
Thank you for the reply, Jordan!

I had `stdlib` compiled with this patch, and got flooded by a number of such warnings for redundant modifiers in extension. Below is the brief statistics, among which 588 are `public`, 2 are `internal`, and the rest 2 are `fileprivate`.

| Source        | Count           |
| ------------- |:-------------:|
| public/SDK/AppKit   | 16   |
| public/SDK/CloudKit  |  2  |
| public/SDK/CoreGraphics | 4  |
| public/SDK/Dispatch | 202  |
| public/SDK/Foundation | 230   |
| public/SDK/Intents | 4  |
| public/SDK/XCTes  | 4   |
| public/core | 130   |
| **TOTAL** | **592**   |

Should all those warnings be fixed before landing this PR?

jrose-apple commented:
Ooh, good thing to check. I should have looked at that first. Yes, we should definitely fix all those, but I can see Dispatch, Foundation, and XCTest all having different preferences from the rest (as the three teams who manage corelibs equivalents for their libraries). @mwwa, @phausler / @millenomi, and @briancroom, would you prefer to standardize on

```
public extension X {
  func foo()
}
```

or

```
extension X {
  public func foo()
}
```

for your respective projects?

phausler commented:
Personally I prefer the latter because it enforces knowingly making things public

dingobye committed:
Fix warnings for stdlib/public/SDK/Foundation

dingobye commented:
Thank you, Philippe! I updated the code to fix the `Foundation` part as per your preference. Let's wait for advice from @mwwa and @briancroom.

mwwa commented:
I'm also happy with @phausler's preference. Seems reasonable, we'll need to perform the equivalent fixes in corelibs-dispatch too.

dingobye committed:
Fix warnings for stdlib/public/SDK/Dispatch

dingobye committed:
Fix warnings for stdlib/public/SDK/XCTest

dingobye committed:
Fix warnings for the rest of stdlib

dingobye commented:
@jrose-apple Hi Jordan, thanks for your all time support, and now all relevant projects have been prepared to embrace this patch. Would you help finalise this, please?

jrose-apple commented:
Ah, I thought we were still waiting on XCTest. I missed that Stuart already took care of that.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
Thanks for all your work here!
